lib/Serialization/SerializedModuleLoader.cpp

```diff
@@ -81,12 +81,42 @@ openModuleFiles(StringRef DirName, StringRef ModuleFilename,
   return std::error_code();
 }
 
+static void addDiagnosticInfoForArchitectureMismatch(ASTContext &ctx,
+                             SourceLoc sourceLocation,
+                             llvm::SmallString<64> moduleName,
+                             llvm::SmallString<16> archName,
+                             llvm::SmallString<128> directoryPath) {
+
+  std::error_code errorCode;
+  llvm::Twine twineDirPath(directoryPath);
+  llvm::sys::fs::directory_iterator directoryIterator(twineDirPath,
+                                                      errorCode,
+                                                      true);
+  llvm::sys::fs::directory_iterator endIterator;
+
+  if(errorCode) {
+      return;
+    }
+
+  std::string foundArchs;
+  for (; directoryIterator != endIterator; directoryIterator.increment(errorCode)) {
+    auto entry = *directoryIterator;
+    llvm::StringRef filePath(entry.path());
+    foundArchs = foundArchs + (foundArchs.length() > 0 ? ", " : "")
+                            + llvm::sys::path::stem(filePath).str();
+  }
+
+  ctx.Diags.diagnose(sourceLocation, diag::sema_no_import_arch,
+                     moduleName, archName, foundArchs);
+}
```

Comment:
@brentdax @CodaFi : I see 2 problems with the current implementation: 

1. 
```cpp
foundArchs = foundArchs + (foundArchs.length() > 0 ? ", " : "")
                        + llvm::sys::path::stem(filePath).str();
```
might not be the most optimal was of appending strings

2. How do I cached the foundArchs for a module ? 

Pretty new to the codebase, so any suggestions will be helpful.